ntoskrnl/ps/quota.c

```diff
@@ -18,123 +17,507 @@ EPROCESS_QUOTA_BLOCK PspDefaultQuotaBlock;
+ * @param[in] Process
+ * A process, represented as a EPROCESS object.
+ * This parameter is used to return the own
+ * process' quota usage.
+ *
+ * @param[in] QuotaBlock
+ * The quota block which quotas are to be returned.
+ * This block can either come from the process itself
+ * or from an object with specified quota charges.
+ *
+ * @param[in] QuotaType
+ * The quota type which quota in question is to
+ * be returned. The permitted types are PsPagedPool,
+ * PsNonPagedPool and PsPageFile.
+ *
+ * @param[in] Amount
+ * The amount of quota to be returned.
+ *
+ * @return
+ * Nothing.
  */
 VOID
 NTAPI
-PspReturnProcessQuotaSpecifiedPool(IN PEPROCESS Process,
-                                   IN PS_QUOTA_TYPE QuotaType,
-                                   IN SIZE_T    Amount)
+PspReturnProcessQuotaSpecifiedPool(
+    _In_opt_ PEPROCESS Process,
+    _In_ PEPROCESS_QUOTA_BLOCK QuotaBlock,
+    _In_ PS_QUOTA_TYPE QuotaType,
+    _In_ SIZE_T Amount)
 {
     KIRQL OldIrql;
-    ASSERT(Process);
-    ASSERT(Process != PsInitialSystemProcess);
+    SIZE_T ReturnThreshold;
+    SIZE_T AmountToReturn = 0;
+    SIZE_T UpdatedQuotaUsage = 0;
+
+    /* Sanity checks */
     ASSERT(QuotaType < PsQuotaTypes);
     ASSERT(!(Amount & 0x80000000)); /* we need to be able to negate it */
 
     /* Guard our quota in a spin lock */
     KeAcquireSpinLock(&PspQuotaLock, &OldIrql);
 
-    if (Process->QuotaUsage[QuotaType] < Amount)
+    /* Does the caller return way more quota than it was previously charged? */
+    if ((Process && Process->QuotaUsage[QuotaType] < Amount) ||
+        QuotaBlock->QuotaEntry[QuotaType].Usage < Amount)
     {
-        DPRINT1("WARNING: Process->QuotaUsage sanity check failed.\n");
+        /* We're returning way too much quota, kill the system! */
+        KeBugCheckEx(QUOTA_UNDERFLOW,
+                     (ULONG_PTR)Process,
+                     (ULONG_PTR)QuotaType,
+                     Process ? (ULONG_PTR)Process->QuotaUsage[QuotaType] :
+                     QuotaBlock->QuotaEntry[QuotaType].Usage,
+                     (ULONG_PTR)Amount);
     }
-    else
+
+    /*
+     * We need to trim the quota limits based on the
+     * amount we're going to return quotas back.
+     */
+    if (QuotaType != PsPageFile &&
+        QuotaBlock != &PspDefaultQuotaBlock)
     {
-        InterlockedExchangeAdd((LONG*)&Process->QuotaUsage[QuotaType],
-                               -(LONG)Amount);
+        /* The return threshold can be non paged or paged */
+        ReturnThreshold = QuotaType ? PSP_NON_PAGED_POOL_QUOTA_THRESHOLD : PSP_PAGED_POOL_QUOTA_THRESHOLD;
+
+        /* Does the amount to return exceeds the threshold? */
+        if (Amount > ReturnThreshold)
+        {
+            /* It does, the new amount is the default */
+            AmountToReturn = ReturnThreshold;
+        }
+        else
+        {
+            /* Otherwise the amount is just the given one by the caller */
+            AmountToReturn = Amount;
+        }
```

Comment:
Looks like a more concise:
```suggestion
        /* If the amount to return exceeds the threshold, the new amount becomes
         * the default, otherwise the amount is just the one given by the caller. */
        AmountToReturn = min(Amount, ReturnThreshold);
```
By the way, the part of the comment that says "the new amount is the default" might somehow need a fix, as the amount being returned is actually the threshold.